src/lib/render-lit-html.ts

```diff
@@ -57,53 +108,183 @@ const getTemplate = (result: TemplateResult) => {
+
+  traverse(ast, {
+    pre(node, parent) {
+      if (isCommentNode(node)) {
+        if (node.data === marker) {
           parts.push({
-            type: 'attribute',
+            type: 'node',
             index: nodeIndex,
-            name,
-            strings,
+          });
+          flushTo(node.sourceCodeLocation!.startOffset);
+          skipTo(node.sourceCodeLocation!.endOffset);
+          ops.push({
+            type: 'node-part',
+            useCustomElementInstance:
+              parent && isElement(parent) && parent.isDefinedCustomElement,
+          });
+          // There will be two comments per part (open+close) in the rendered
+          // output and on the client, so increment again for that
+          nodeIndex++;
+        }
+      } else if (isElement(node)) {
+        // Whether to flush the start tag. This is neccessary if we're changing
+        // any of the attributes in the tag, so it's true for custom-elements
+        // which might reflect their own state, or any element with a binding.
+        let writeTag = false;
+        let boundAttrsCount = 0;
+
+        const tagName = node.tagName;
+        let ctor;
+
+        if (tagName.indexOf('-') !== -1) {
+          // Looking up the constructor here means that custom elements must be
+          // registered before rendering the first template that contains them.
+          ctor = customElements.get(tagName);
+          if (ctor !== undefined) {
+            // Write the start tag
+            writeTag = true;
+            // Mark that this is a custom element
+            node.isDefinedCustomElement = true;
+            ops.push({
+              type: 'custom-element-open',
+              tagName,
+              ctor,
+            });
+          }
+        }
+        if (node.attrs.length > 0) {
+          for (const attr of node.attrs) {
+            if (attr.name.endsWith(boundAttributeSuffix)) {
+              writeTag = true;
+              boundAttrsCount += 1;
+              // Note that although we emit a lit-bindings comment marker for any
+              // nodes with bindings, we don't account for it in the nodeIndex because
+              // that will not be injected into the client template
+              const name = attr.name.substring(
+                0,
+                attr.name.length - boundAttributeSuffix.length
+              );
+              const strings = attr.value.split(markerRegex);
+              const attrSourceLocation = node.sourceCodeLocation!.attrs[
+                attr.name
+              ];
+              const attrNameStartOffset = attrSourceLocation.startOffset;
+              const attrEndOffset = attrSourceLocation.endOffset;
+              parts.push({
+                type: 'attribute',
+                index: nodeIndex,
+                name,
+                strings,
+              });
+              flushTo(attrNameStartOffset);
+              ops.push({
```

Comment:
Maybe just a comment here or where this opcode is consumed, but it would be good to just note that the "attribute-part" opcode corresponds to an "AttributeCommitter" (group of one or more values being set to a single attribute) on the instance-side.  This always trips me up, so would be nice to clarify somehow when we're talking about the grouped-type of part or the per-value type of part.

Reply:
done